crates/ty_python_semantic/src/ast_node_ref.rs

```diff
@@ -32,54 +32,85 @@ use ruff_db::parsed::ParsedModuleRef;
 /// run on every AST change. All other queries only run when the expression's identity changes.
 #[derive(Clone)]
 pub struct AstNodeRef<T> {
-    /// Owned reference to the node's [`ParsedModuleRef`].
-    ///
-    /// The node's reference is guaranteed to remain valid as long as it's enclosing
-    /// [`ParsedModuleRef`] is alive.
-    parsed: ParsedModuleRef,
+    /// A pointer to the [`ruff_db::parsed::ParsedModule`] that this node was created from.
+    module_ptr: *const (),
+
+    /// Debug information.
+    #[cfg(debug_assertions)]
+    kind: ruff_python_ast::NodeKind,
+    #[cfg(debug_assertions)]
+    range: ruff_text_size::TextRange,
 
-    /// Pointer to the referenced node.
-    node: std::ptr::NonNull<T>,
+    /// The index of the node in the AST.
+    index: NodeIndex,
+
+    _node: PhantomData<T>,
 }
 
-#[expect(unsafe_code)]
-impl<T> AstNodeRef<T> {
-    /// Creates a new `AstNodeRef` that references `node`. The `parsed` is the [`ParsedModuleRef`] to
-    /// which the `AstNodeRef` belongs.
-    ///
-    /// ## Safety
+impl<T> AstNodeRef<T>
+where
+    T: HasNodeIndex + Ranged + PartialEq + Debug,
+    for<'ast> AnyNodeRef<'ast>: From<&'ast T>,
+    for<'ast> &'ast T: TryFrom<AnyRootNodeRef<'ast>>,
+{
+    /// Creates a new `AstNodeRef` that references `node`.
     ///
-    /// Dereferencing the `node` can result in undefined behavior if `parsed` isn't the
-    /// [`ParsedModuleRef`] to which `node` belongs. It's the caller's responsibility to ensure that
-    /// the invariant `node belongs to parsed` is upheld.
-    pub(super) unsafe fn new(parsed: ParsedModuleRef, node: &T) -> Self {
+    /// This method may panic or produce unspecified results if the provided module is from a
+    /// different file or Salsa revision than the module to which the node belongs.
+    pub(super) fn new(module_ref: &ParsedModuleRef, node: &T) -> Self {
+        debug_assert_eq!(
+            module_ref.get_by_index(node.node_index()).try_into().ok(),
+            Some(node)
+        );
+
         Self {
-            parsed,
-            node: std::ptr::NonNull::from(node),
+            module_ptr: module_ref.module().as_ptr(),
+            #[cfg(debug_assertions)]
+            kind: AnyNodeRef::from(node).kind(),
+            #[cfg(debug_assertions)]
+            range: node.range(),
+            index: node.node_index().clone(),
```

Comment:
Nit: I still think it would be nice to have a non atomic `NodeIndex`, because it isn't really required here